src/trace-plugin-loader.ts

```diff
@@ -16,278 +16,485 @@
+   * @param version The module version.
+   */
+  applyPlugin<T>(moduleExports: T, file: string, version: string): T;
 }
 
-function isInternalPatch<T>(obj: InternalInstrumentation<T>):
-    obj is InternalPatch<T> {
-  return !!(obj as InternalPatch<T>).patch;
-}
+/**
+ * A class that represents wrapper logic around a user-provided plugin object
+ * to be applied to a single module.
+ */
+export class ModulePluginWrapper implements PluginWrapper {
+  // Sentinel value to indicate that a plugin has not been loaded into memory
+  // yet.
+  private static readonly NOT_LOADED: Plugin = [];
+  private unpatchFns: Array<() => void> = [];
+  // A logger.
+  private logger: Logger;
+  // Configuration for a TraceAgent instance.
+  private traceConfig: TraceAgentConfig;
+  // Display-friendly name of the module being patched by this plugin.
+  private name: string;
+  // The path to the plugin.
+  private path: string;
+  // The exported value of the plugin, or NOT_LOADED if it hasn't been
+  // loaded yet.
+  private pluginExportedValue: Plugin = ModulePluginWrapper.NOT_LOADED;
+  private traceApiInstances: TraceAgent[] = [];
 
-function isInternalIntercept<T>(obj: InternalInstrumentation<T>):
-    obj is InternalIntercept<T> {
-  return !!(obj as InternalIntercept<T>).intercept;
-}
+  /**
+   * Constructs a new PluginWrapper instance.
+   * @param logger The logger to use.
+   * @param options Initialization fields for this object.
+   * @param traceConfig Configuration for a TraceAgent instance.
+   */
+  constructor(
+      logger: Logger, options: ModulePluginWrapperOptions,
+      traceConfig: TraceAgentConfig) {
+    this.logger = logger;
+    this.name = options.name;
+    this.path = options.path;
+    this.traceConfig = traceConfig;
+  }
 
-let plugins: PluginStore = Object.create(null);
-// tslint:disable-next-line:no-any
-let intercepts: {[moduleName: string]: {interceptedValue: any}} =
-    Object.create(null);
-let activated = false;
-// TODO(kjin): Make plugin loader a singleton, so we can avoid shadowing.
-// tslint:disable-next-line:variable-name
-let logger_: Logger;
-
-function checkLoadedModules(): void {
-  for (const moduleName of Object.keys(plugins)) {
-    // \\ is benign on unix and escapes \\ on windows
-    const regex =
-        new RegExp('node_modules\\' + path.sep + moduleName + '\\' + path.sep);
-    for (const file of Object.keys(require.cache)) {
-      if (file.match(regex)) {
-        logger_.error(
-            moduleName + ' tracing might not work as ' + file +
-            ' was loaded before the trace agent was initialized.');
-        break;
+  isSupported(version: string): boolean {
+    // The plugin is lazily loaded here.
+    const plugin = this.getPluginExportedValue();
+    // Count the number of Patch/Intercept objects with compatible version
+    // ranges
+    let numFiles = 0;
+    plugin.forEach(instrumentation => {
+      const postLoadVersions = instrumentation.versions;
```

Comment:
I'm confused, Why is this variable named *post*LoadVersions?

Reply:
> Just something to keep in mind if you see something weird like storing plugins as an array or two layers of version-checks (you should make a note of it if it doesn't make sense).

This is exactly the kind of artifact I was describing in the PR abstract. There used to be pre-load and post-load versions. Thanks for the catch!